owb_rmt.c

```diff
@@ -0,0 +1,441 @@
+                        {
+                            // rising edge occured before 15us -> bit 1
+                            read_data |= 0x80;
+                        }
+                    }
+                }
+
+                read_data >>= 8 - number_of_bits_to_read;
+            }
+
+            vRingbufferReturnItem( info->rb, (void *)rx_items );
+        } else
+        {
+            ESP_LOGE(TAG, "%s(): rx_items == 0", __func__);
+
+            // time out occurred, this indicates an unconnected / misconfigured bus
+            res = OWB_STATUS_HW_ERROR;
+        }
+    } else {
+        // error in tx channel
+        ESP_LOGE(TAG, "Error tx");
+        res = OWB_STATUS_HW_ERROR;
+    }
+
+    rmt_rx_stop( info->rx_channel );
+
+    *in = read_data;
+    return res;
+}
+
+static owb_status _uninitialize(const OneWireBus *bus)
+{
+    owb_rmt_driver_info *info = info_of_driver(bus);
+
+    rmt_driver_uninstall( info->tx_channel );
+    rmt_driver_uninstall( info->rx_channel );
+
+    return OWB_STATUS_OK;
+}
+
+static struct owb_driver rmt_function_table =
+{
+    .name = "owb_rmt",
+    .uninitialize = _uninitialize,
+    .reset = _reset,
+    .write_bits = _write_bits,
+    .read_bits = _read_bits
+};
+
+static owb_status _init( owb_rmt_driver_info *info, uint8_t gpio_num,
+                         rmt_channel_t tx_channel, rmt_channel_t rx_channel )
+{
+    owb_status status = OWB_STATUS_HW_ERROR;
+
+    info->bus.driver = &rmt_function_table;
+    info->tx_channel = tx_channel;
+    info->rx_channel = rx_channel;
+    info->gpio = gpio_num;
+
+#ifdef OW_DEBUG
+    ESP_LOGI(TAG, "RMT TX channel: %d", info->tx_channel);
+    ESP_LOGI(TAG, "RMT RX channel: %d", info->rx_channel);
+#endif
+
+    rmt_config_t rmt_tx;
+    rmt_tx.channel = info->tx_channel;
+    rmt_tx.gpio_num = gpio_num;
+    rmt_tx.mem_block_num = 1;
+    rmt_tx.clk_div = 80;
+    rmt_tx.tx_config.loop_en = false;
+    rmt_tx.tx_config.carrier_en = false;
+    rmt_tx.tx_config.idle_level = 1;
+    rmt_tx.tx_config.idle_output_en = true;
+    rmt_tx.rmt_mode = RMT_MODE_TX;
+    if(rmt_config(&rmt_tx) == ESP_OK)
+    {
+        if(rmt_driver_install(rmt_tx.channel, 0, ESP_INTR_FLAG_LOWMED | ESP_INTR_FLAG_IRAM | ESP_INTR_FLAG_SHARED) == ESP_OK)
+        {
+            rmt_config_t rmt_rx;
+            rmt_rx.channel = info->rx_channel;
+            rmt_rx.gpio_num = gpio_num;
+            rmt_rx.clk_div = 80;
+            rmt_rx.mem_block_num = 1;
+            rmt_rx.rmt_mode = RMT_MODE_RX;
+            rmt_rx.rx_config.filter_en = true;
+            rmt_rx.rx_config.filter_ticks_thresh = 30;
+            rmt_rx.rx_config.idle_threshold = OW_DURATION_RX_IDLE;
+            if (rmt_config( &rmt_rx ) == ESP_OK)
+            {
+                if (rmt_driver_install( rmt_rx.channel, 512, ESP_INTR_FLAG_LOWMED | ESP_INTR_FLAG_IRAM | ESP_INTR_FLAG_SHARED ) == ESP_OK)
+                {
+                    rmt_get_ringbuf_handle( info->rx_channel, &info->rb );
```

Comment:
@DavidAntliff 

I was hoping we could retain the new externally determined static/dynamic approach. I removed the allocation/free routines because the end user could simply malloc the structure if they'd like, or statically allocate it. It didn't seem necessary to duplicate functionality that is built into the language, it just made the library more complex. It confused me the first time I used it actually.

The PIN_INPUT_ENABLE exists in the latest esp-idf and is defined in a rom header include file in esp-idf. I found I had to use those routines directly as the gpio driver routines reconfigured the pins into gpio mode, disabling rmt mode.

I see there is an IDF_VER variable and a esp_get_idf_version() but I don't see an way to tell the version at compile time. We could always ask or document that you need latest to run with the RMT. I'm guessing 3.0 will release relatively soon.

Reply:
@chmorgan I hear you. I can see how the interface I had might be confusing. If it helps, this is what I'm trying to do:

https://github.com/DavidAntliff/esp32-components/wiki/Component-Design

In my experience with building large commercial apps with C, providing external allocation can cause problems with encapsulation and future API changes. I've used heavier techniques in other code bases so I'm trying to create a more lightweight solution that still makes some sense. I admit it's not perfect and I'm still trying to find a good compromise.

Fair point on the `PIN_INPUT_ENABLE` macro - if it works (and it does) then that's fine for now.

I'm not sure when 3.0 will be released - it's been "pending" for many months. However we could define a macro that can be easily changed (and commented) between API versions. Then the main line can be set for 2.x, and you can easily use a compiler directive to override it in your "pre-3.0" build.

Reply:
@DavidAntliff I can appreciate the consistent approach with the modules. I'm sure you noticed but to handle the instantiate/cleanup along with api and internal structure changes I created driver specific init function and a generic cleanup function. It's your repository and your choice on how you'd like it to look in the end!

Not sure when 3.0 will release but maybe a compile time define would work for that. Then users of the older or newer esp-idf (I'd probably assume the newer esp-idf by default but again your choice) can set the appropriate define in their build.

Reply:
Ok, in order to move forward, how about I merge this PR (and also merge the changes in esp32-ds18b20 and esp32-ds18b20-example I have made to match the API changes ), on the understanding that I'd like to revisit the API myself, so I make no guarantees that it might not change in the near or distant future? After that, if it's not quite how you like it, then you could either wrap it, or even fork if you wish. Sound OK? :)

I'll probably also add a compile-time define some time after the merge for selecting between the 2.x or future, but for now I'm happy to leave it building with "pre-3.0" since that's the future of ESP-IDF (at least as far as we know at this time).